command/job_restart.go

```diff
@@ -0,0 +1,955 @@
+		if _, ok := alloc.TaskStates[task]; !ok {
+			continue
+		}
+
+		c.ui.Output(fmt.Sprintf(
+			"    %s: Restarting task %q in allocation %q for group %q",
+			formatTime(time.Now()),
+			task,
+			shortAllocID,
+			alloc.TaskGroup,
+		))
+
+		restarts.Go(func(task string) func() error {
+			return func() error {
+				err := c.client.Allocations().Restart(alloc, task, nil)
+				if err != nil {
+					return fmt.Errorf("Failed to restart task %q: %v", task, err)
+				}
+				return nil
+			}
+		}(task))
+	}
+	return restarts.Wait().ErrorOrNil()
+}
+
+// stopAlloc stops an allocation and blocks until the replacement allocation is
+// running.
+func (c *JobRestartCommand) stopAlloc(alloc *api.Allocation) error {
+	shortAllocID := limit(alloc.ID, c.length)
+
+	c.ui.Output(fmt.Sprintf(
+		"    %s: Rescheduling allocation %q for group %q",
+		formatTime(time.Now()),
+		shortAllocID,
+		alloc.TaskGroup,
+	))
+
+	var q *api.QueryOptions
+	if c.noShutdownDelay {
+		q = &api.QueryOptions{
+			Params: map[string]string{"no_shutdown_delay": "true"},
+		}
+	}
+
+	// Stop allocation and wait for its replacement to be running or for a
+	// blocked evaluation that prevents placements for this task group to
+	// happen.
+	_, err := c.client.Allocations().Stop(alloc, q)
+	if err != nil {
+		return fmt.Errorf("Failed to stop allocation %q: %v", shortAllocID, err)
+	}
+
+	// errCh receives an error if anything goes wrong or nil when the
+	// replacement allocation is running.
+	// Use a buffered channel to prevent both goroutine from blocking trying to
+	// send a result back.
+	errCh := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go c.monitorBlockedEvals(ctx, alloc, errCh)
+	go c.monitorReplacementAlloc(ctx, alloc, errCh)
+
+	// If we receive an error and nil it's safe to ignore the error since the
+	// nil result is what we are looking for.
+	return <-errCh
+}
+
+// monitorBlockedEvals searches for blocked evaluations for the allocation job.
+//
+// Returns an error in errCh if anything goes wrong or if there are placement
+// failures for the allocation task group.
+func (c *JobRestartCommand) monitorBlockedEvals(ctx context.Context, alloc *api.Allocation, errCh chan<- error) {
+	q := &api.QueryOptions{WaitIndex: 1}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		evals, qm, err := c.client.Jobs().Evaluations(alloc.JobID, q)
+		if err != nil {
+			errCh <- fmt.Errorf("Failed to retrieve evaluations for job %q: %v", alloc.JobID, err)
+			return
+		}
+
+		for _, eval := range evals {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			if eval.Status != api.EvalStatusBlocked {
+				continue
+			}
+
+			failures := eval.FailedTGAllocs[alloc.TaskGroup]
+			if failures != nil {
+				errCh <- fmt.Errorf("%s:\n%s",
+					jobRestartFailedToPlaceNewAllocation,
+					formatAllocMetrics(failures, false, strings.Repeat(" ", 4)),
+				)
+				return
+			}
```

Comment:
This will throw an error and stop the restart if there are _any_ blocked evaluations for the job, including ones that were for unrelated issues. Consider the following scenario:

* `nomad job run`
* 5 allocs placed and running, 1 alloc queued (which means we have 1 blocked eval)
* `nomad job restart -reschedule`
* 1 running alloc is stopped, and the restart command exits here

We might be able to make this work if we used the wait index we got back from the initial stop, so that we're only detecting _new_ blocked evaluations that happened after the stop. That still wouldn't account for random other allocations that crash and fail to get placed. But I think that's an ok gap to have -- if that starts happening you probably want to stop what you're doing anyways.

Reply:
> We might be able to make this work if we used the wait index we got back from the initial stop, so that we're only detecting new blocked evaluations that happened after the stop.

I had this in my first implementation, but that's how I found out that `api.Allocations().Stop()` doesn't return an index 😅 

I will fix the `api` package in a separate PR and merge it here so I can use the right index.

> That still wouldn't account for random other allocations that crash and fail to get placed. But I think that's an ok gap to have -- if that starts happening you probably want to stop what you're doing anyways.

Yeah, I thought about that as well, but couldn't come up with a good solution. Ideally the eval broker would make sure the eval chain is kept, but I found some situations where the eval doesn't reference any follow-up (like in [`MarkForCancel`](https://github.com/hashicorp/nomad/blob/v1.5.0/nomad/eval_broker.go#L1015-L1032)). I tried fixing some of those, but I was still getting some evals breaking the chain, so checking for _any_ blocked eval with placement failures was the best I could think of.

Stopping allocs while there are placement failures is problematic because it can leave you with zero allocations running, so I thought it would be better to stop the process.

Reply:
Agreed :+1:

Reply:
I implemented this in https://github.com/hashicorp/nomad/pull/16278/commits/9552e091678a2e07aaac99d71dbd620e78e339ef but then I noticed that the blocking query index only controls when the request is unblocked, not the data returned, so we still receive the full list of evaluations.

I fixed this in https://github.com/hashicorp/nomad/pull/16278/commits/b0cc1234d39300bc7fcacf03d8188ed6e21d61cb by skipping evaluations crated before the job stop index.

I also had to remove the check for [`eval.Status != api.EvalStatusBlocked`](https://github.com/hashicorp/nomad/pull/16278/commits/9552e091678a2e07aaac99d71dbd620e78e339ef#diff-331d4a10a3b907d64a5d41fda14da628e6b34efd7f7a0095e7ba52782279ff10L729-L731) because when a new placement failure happens the new eval is cancelled with description `existing blocked evaluation exists for job "example"` so it never gets blocked.